### PR TITLE
Enable tor LnUrls

### PR DESCRIPTION
**Pull request by dannydeezy**

### CODE REVIEW

**Comment by dannydeezy** [comment]
- [ ] needs to be tested

**Comment by dannydeezy** [comment]
for https://github.com/dannydeezy/deezy-auto-earn/issues/2

**Review by aphex3k — COMMENTED** [review]
Love to see it 👍🏻

**Comment by aphex3k** [comment]
```
(node:8) [DEP0123] DeprecationWarning: Setting the TLS ServerName to an IP address is not permitted by RFC 6066. This will be ignored in a future version.
(Use `node --trace-deprecation ...` to show where the warning was created)
Already connected to deezy
Fetching channel info
Found 1 locally initiated channel(s) with deezy
Checking if any deezy channels are ready to close
Checking if we should open a channel to deezy
Total local open or pending sats: 1471363
Not opening channel, total local sats 1471363 is above threshold 1000000
ReferenceError: source is not defined
    at Object.fetchInvoice (/deezy-auto-earn/clients/lnurl.js:19:36)
    at attemptPaymentToDestination (/deezy-auto-earn/index.js:49:41)
    at run (/deezy-auto-earn/index.js:208:15)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
ReferenceError: source is not defined
    at Object.fetchInvoice (/deezy-auto-earn/clients/lnurl.js:19:36)
    at attemptPaymentToDestination (/deezy-auto-earn/index.js:49:41)
    at run (/deezy-auto-earn/index.js:208:15)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
ReferenceError: source is not defined
    at Object.fetchInvoice (/deezy-auto-earn/clients/lnurl.js:19:36)
    at attemptPaymentToDestination (/deezy-auto-earn/index.js:49:41)
    at run (/deezy-auto-earn/index.js:208:15)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
ReferenceError: source is not defined
    at Object.fetchInvoice (/deezy-auto-earn/clients/lnurl.js:19:36)
    at attemptPaymentToDestination (/deezy-auto-earn/index.js:49:41)
    at run (/deezy-auto-earn/index.js:208:15)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
ReferenceError: source is not defined
    at Object.fetchInvoice (/deezy-auto-earn/clients/lnurl.js:19:36)
    at attemptPaymentToDestination (/deezy-auto-earn/index.js:49:41)
    at run (/deezy-auto-earn/index.js:208:15)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```
I have configured different LNURLs with onion addresses but I get the same error for each of them.

Is this an issue with the LNURLp or with the code?

image

Thunderhub is able to decode it correctly

**Comment by aphex3k** [comment]
Let me try again. It might be not connecting to tor proxy correctly...

File: clients/lnurl.js [hunk 1]
```diff
@@ -1,12 +1,31 @@
 
-const { requestInvoice } = require('lnurl-pay')
+const { requestInvoice, utils } = require('lnurl-pay')
+const { isOnionUrl, decodeUrlOrAddress } = utils
+const tor_axios = require('tor-axios');
+const config = require('../config.json')
+
+const tor = tor_axios.torSetup({
+    ip: 'localhost',
```
Comment (aphex3k) [inline clients/lnurl.js]: Can we make `localhost` configurable? In Umbrel, tor is never running on localhost... Just like the tor port is configurable, make the host/ip configurable as well :D

**Review by aphex3k — CHANGES_REQUESTED** [review]

File: clients/lnurl.js (hunk 1 above)
Comment (dannydeezy) [reply clients/lnurl.js]: updated

**Comment by aphex3k** [comment]
Unfortunately Im still seeing the same error as mentioned above :( 

```
LNURL1DP68GUP69UHK6CT4VD4K26N9D4EX7M3JXEKHSEMDXA4NVENDDENHSER5XEUKYME5VVMHWVMTX4UXGUMVD35H5VNWV4KHJUMVWD5KGTN0DE5K7M30D3H82UNVWQHKZURF9AMRZTMVDE6HYMP0XCSTA7HE
```

Above is an LNURLp for 50 sats to one of the onion hosts I use. Anyone should be able to use that for testing purposes

File: clients/lnurl.js [hunk 2]
```diff
@@ -1,12 +1,31 @@
...
+})
+
+async function torGet({ url, params }) {
+    console.log({ url, params })
+    const resp = await tor.get(url, { params })
+    return resp.data
+}
 
 async function fetchInvoice({ lnUrlOrAddress, paymentAmountSats }) {
+    const url = decodeUrlOrAddress(source.lnurl_or_lnaddress)
```
Comment (aphex3k) [inline clients/lnurl.js]: Should this read instead?
```
    const url = decodeUrlOrAddress(lnUrlOrAddress)
```
I think `source` is what is throwing my error...

> ReferenceError: source is not defined

`source` in the error message is probably referring to the variable name `source` in this line but it is not defined. The argument to the function is actually `lnUrlOrAddress` instead.

**Review by aphex3k — CHANGES_REQUESTED** [review]

File: clients/lnurl.js (hunk 2 above)
Comment (dannydeezy) [reply clients/lnurl.js]: yup good catch

Comment (dannydeezy) [reply clients/lnurl.js]: the dangers of not having any tests!

**Review by aphex3k — APPROVED** [review]
SUCCESS!
```
(node:8) [DEP0123] DeprecationWarning: Setting the TLS ServerName to an IP address is not permitted by RFC 6066. This will be ignored in a future version.
(Use `node --trace-deprecation ...` to show where the warning was created)
Already connected to deezy
Fetching channel info
Found 1 locally initiated channel(s) with deezy
Checking if any deezy channels are ready to close
Checking if we should open a channel to deezy
Total local open or pending sats: 603288
Chain balance is 361141
Not opening channel, chain balance 361141 is below threshold 1500000 plus buffer 50000
http://mauckejemron26mxgm7k6fmngxdt6ybo4c7w3k5xdslliz2nemyslsid.onion/lnurlp/api/v1/lnurl/4
Fetching invoice from LNURL1DP68GUP69UHK6CT4VD4K26N9D4EX7M3JXEKHSEMDXA4NVENDDENHSER5XEUKYME5VVMHWVMTX4UXGUMVD35H5VNWV4KHJUMVWD5KGTN0DE5K7M30D3H82UNVWQHKZURF9AMRZTMVDE6HYMP0XSWLEJFY for 425000 sats
{
  url: 'http://mauckejemron26mxgm7k6fmngxdt6ybo4c7w3k5xdslliz2nemyslsid.onion/lnurlp/api/v1/lnurl/4',
  params: undefined
}
{
  url: 'http://mauckejemron26mxgm7k6fmngxdt6ybo4c7w3k5xdslliz2nemyslsid.onion:80/lnurlp/api/v1/lnurl/cb/4',
  params: { amount: 425000000 }
}
Using max route fee sats 4
Payment confirmed, with fee 2 satoshis, and ppm 5
```
